### PR TITLE
Bug 2037075: Verify Builds with CSI Volume Sources

### PR DESCRIPTION
**Pull request by jkhelil**
https://issues.redhat.com/browse/BUILD-275

### CODE REVIEW

File: test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml
Comment (yselkowitz) [inline test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:18]: Please note that this image is being relocated for multi-arch support in #26598

Comment (jkhelil) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:18]: ack @yselkowitz

**Comment by gabemontero** [comment]
you have to run `make update` to update the bindata @jkhelil 

you also minimally have to assign @adambkaplan to this PR to get teh approved label per https://github.com/openshift/origin/pull/26646#issuecomment-981671576

File: test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml
Comment (gabemontero) [inline test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: Per my comment to you in https://docs.google.com/document/d/1F3VFkiLUJbsUkFJcjnA2oyROYTI5RRlTJYl4ZWQoNls/edit?disco=AAAASUDEeYE add the `‘refreshResource’ attribute, and set it to a string value of false

Comment (gabemontero) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: Another option as well, which I might like better, is that in your OCM controller PR, if you see CSI driver volume mounts with our shared resource driver here, you adjust their volume attributes to force `refreshResources` to 'false'

WDYT @adambkaplan

File: test/extended/testdata/builds/volumes/csi-s2i-buildconfig.yaml
Comment (gabemontero) [inline test/extended/testdata/builds/volumes/csi-s2i-buildconfig.yaml:33]: same refreshResource to false comment

File: test/extended/testdata/builds/volumes/csi-sharedresourcerolebinding.yaml
Comment (gabemontero) [inline test/extended/testdata/builds/volumes/csi-sharedresourcerolebinding.yaml:11]: builds use the 'builder' service account  ... change default to builder

**Review by gabemontero — CHANGES_REQUESTED** [review]
I'm guessing the ginkgo code of the e2e's is just following the pattern @coreydaley established with his original build volume work 

assign this PR to him as well so he can review these

otherwise I added some shared resource specific comments, along with a question to @adambkaplan that might mean a tweak to what you are doing in your OCM PR

File: test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml
Comment (jkhelil) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: @gabemontero I added the refreshResource attribute

File: test/extended/testdata/builds/volumes/csi-s2i-buildconfig.yaml
Comment (jkhelil) [reply test/extended/testdata/builds/volumes/csi-s2i-buildconfig.yaml:33]: I added the refreshResource attribute

File: test/extended/testdata/builds/volumes/csi-sharedresourcerolebinding.yaml
Comment (jkhelil) [reply test/extended/testdata/builds/volumes/csi-sharedresourcerolebinding.yaml:11]: Updated

File: test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml
Comment (gabemontero) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: OK putting a comment to reflect our team discussion yesterday on this after I posted my last set of comments, namely
we are *NOT* going to have the OCM force shared resource CSI volumes to have refreshResource set to false.

That said, it is a supported option we do want to let user opt into.

So I'll need to adjust my request to you here @jkhelil ..... for each of your shared resource test items here, have one where you do not set refreshResource, and then one where you add the refreshsResource: 'false' setting

**Comment by gabemontero** [comment]
/assign @gabemontero 

/assign @coreydaley 
(to minimally review your volume test cases)

/assign @adambkaplan 
(to minimally get the approve label)

see if you can remember @jkhelil to start doing these assigns yourself.

**Comment by jkhelil** [comment]
ack @gabemontero , I do remember, Just not clear for me if I should wait until all tests are green to assign the PR for approval or not

**Comment by gabemontero** [comment]
> ack @gabemontero , I do remember, Just not clear for me if I should wait until all tests are green to assign the PR for approval or not

I would say if there is more than likely that the failed test are non-related flakes, go ahead and assign / start the review process.

File: test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml
Comment (jkhelil) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: @gabemontero what is the purpose for adding tests without refreshsResource: 'false', because we know that it shouldnt happen for builds and it will fail(correct me if i am wring about it)

Comment (gabemontero) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: no @jkhelil both scenarios should work
- if the user does not add the `refreshResource` volume attribute, it should work
- if the user does add the `refreshResourece: 'false'` volume attribute, it should work

Hence, I want you to make sure both paths work but having 2 flavors of your various tests here.

Comment (jkhelil) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: @gabemontero can you have a look please  to the changes regarding this last comment

Comment (gabemontero) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: ok I'm good with the mix you have for this @jkhelil

Comment (gabemontero) [reply test/extended/testdata/builds/volumes/csi-docker-buildconfig.yaml:33]: git the approve label from @adambkaplan and see if @coreydaley is OK with the way you added to the volumes tests he did, and we'll go from there

also, if the agnostic-cmd test is the only failure in the tests, and if you examine the tests and provide a comment to use explaining that all the sig-builds test have passed, and none of any remaining flakes are build related, we can use the skip prow command to bypass that one

**Review by adambkaplan — CHANGES_REQUESTED** [review]
@jkhelil we should also have a test on a standard cluster which tries to mount a CSI volume. In this test case, we would expect the build to be created, and then later fail.

**Comment by adambkaplan** [comment]
/retitle Bug 2037075: Verify Builds with CSI Volume Sources

**Comment by jkhelil** [comment]
retrying test flaking, failure is not related to the PR
/test e2e-gcp-upgrade

**Comment by gabemontero** [comment]
sig-builds in the last e2e-aws-single-node jobs all passed, and the flake was unrelated

since that test appears optional (does not have the required tag in tide) going to see if we can skip it

/skip

**Comment by gabemontero** [comment]
yep appears so  ... down to the upgrade job

/retest

**Comment by gabemontero** [comment]
> @jkhelil we should also have a test on a standard cluster which tries to mount a CSI volume. In this test case, we would expect the build to be created, and then later fail.

@jkhelil I still don't see where you added a test of the type @adambkaplan called for above ^^

with such a test, I would expect a tweak in your `JustBeforeEach` logic where you do not skip the test when it is on a non-tech preview cluster and you want to verify that the build fails .... or you may have to move that skip test outside of the `JustBeforeEach` depending on how you can or cannot parameterize that block.

**Comment by jkhelil** [comment]
> > @jkhelil we should also have a test on a standard cluster which tries to mount a CSI volume. In this test case, we would expect the build to be created, and then later fail.
> 
> @jkhelil I still don't see where you added a test of the type @adambkaplan called for above ^^
> 
> with such a test, I would expect a tweak in your `JustBeforeEach` logic where you do not skip the test when it is on a non-tech preview cluster and you want to verify that the build fails .... or you may have to move that skip test outside of the `JustBeforeEach` depending on how you can or cannot parameterize that block.

**Review by gabemontero — APPROVED** [review]
works for me

/lgtm

still minimally need an approve from @adambkaplan I believe

File: test/extended/builds/volumes.go
Comment (adambkaplan) [inline test/extended/builds/volumes.go:179]: I _don't_ think this will work the way you expect it to. I'm pretty positive this breaks how openshift-tests works with ginkgo.

I would use skips in each test instead, and/or separate context/Describe sections for tech preview vs. standard cluster.

See https://github.com/openshift/origin/blob/master/test/extended/README.md

File: test/extended/builds/volumes.go
Comment (adambkaplan) [inline test/extended/builds/volumes.go:350]: Be explicit that you're checking for `TechPreviewNoUpgrade`. There are other feature sets that end with `NoUpgrade`.

```suggestion
// isTechPreviewNoUpgrade checks if a cluster is a TechPreviewNoUpgrade cluster
func isTechPreviewNoUpgrade(config *rest.Config) (bool, error) {
```

File: test/extended/builds/volumes.go
Comment (adambkaplan) [inline test/extended/builds/volumes.go:351]: I don't think you need to use the dynamic client. The test framework includes typed clients for all OpenShift APIs.

File: test/extended/builds/volumes.go
Comment (jkhelil) [reply test/extended/builds/volumes.go:350]: fixed

File: test/extended/builds/volumes.go
Comment (jkhelil) [reply test/extended/builds/volumes.go]: fixed

File: test/extended/builds/volumes.go
Comment (jkhelil) [reply test/extended/builds/volumes.go]: Thank you @adambkaplan , Good catch
Fixed now

**Comment by gabemontero** [comment]
bump @adambkaplan 

are you at least willing to apply the approve label to @jkhelil PR and have others complete the review, or do you think you'll be able to get lgtm level cycles to look at this soon.

other than non related flakes in the agnostic cmd suite,  this appears green

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Pull request merged**
